### PR TITLE
OS installation hotfix.

### PR DESCRIPTION
**Pull request by JuneZhou19**
1. fix for max retry of discover workflow check completion.
2. fix not able to get workflow instance id when os install already finished upon verify os installation workflow status check.

@InfraSIM/infrasim_dev

### CODE REVIEW

**Comment by JuneZhou19** [comment]
Pls help review. It is going to fix failure below from jenkins run for OS installation. @turtle-fly @XiaowenJiang @chenge3 @fub2 

02:59:18  06:57:04.851                puffer:  Test case failed: 
02:59:18  [Summary]
02:59:18  **1. fail**
02:59:18  
02:59:18  Unexpected exception happened during case running, trace back: 
02:59:18  Traceback (most recent call last):
02:59:18    File "/home/infrasim/jenkins/workspace/docker-rackhd/test/case/CBaseCase.py", line 337, in run
02:59:18      self.test()
02:59:18    File "/home/infrasim/jenkins/workspace/docker-rackhd/test/case/interactive/T0000_rackhd_OSInstall.py", line 136, in test
02:59:18      rsp = node.get_workflow_by_id(workflow_instance_id)
02:59:18  UnboundLocalError: local variable 'workflow_instance_id' referenced before assignment

File: case/interactive/T0000_rackhd_OSInstall.py [hunk 1]
```diff
@@ -100,8 +100,11 @@ def test(self):
                             self.log("WARNING", "There is {0} active workflow(s) on node {1}."
                                                 "Will not continue with OS installtion on this node\n".
                                         format(len(rsp), node_id))
-                            node_info = (node_id, node)
-                            node_os_install_list.remove(node_info)
+                            index = 0
+                            for node_id_pend, node_pend in node_os_install_list:
+                                if node_id_pend == node_id:
+                                    node_os_install_list.pop(index)
+                                index = index + 1
```
Comment (turtle-fly) [inline case/interactive/T0000_rackhd_OSInstall.py]: This is a node traverse inside a node traverse, which is not good.
I believe what you want to do is to delete this key-value pair from the dict: `del(node_os_install_list[node_id])`, considering the probable key missing (which will not in this test), you can try...except to protect the code:

    try:
        del(node_os_install_list[node_id])
    except KeyError:
        pass

P.S the node_os_install_list is actually a dict, right?

File: case/interactive/T0000_rackhd_OSInstall.py [hunk 2]
```diff
@@ -114,13 +117,21 @@ def test(self):
...
             self.monorail.put_node_ipmi_obm(
                 node_id=node_id, host=node_bmc_ip, username="admin", password="admin")
             node.install_os(self.data["os_name"])
             self.log(
                 'WARNING', 'Installation on node {} starts, please open VNC to check'.format(node_id))
+            rsp = node.get_workflows(active=True).items()
+            if len(rsp) != 0:
+                workflow_obj = rsp[0][1]
+                workflow_instance_id = workflow_obj.instanceId
+                node_workflow_dic[node_id] = workflow_instance_id
```
Comment (turtle-fly) [inline case/interactive/T0000_rackhd_OSInstall.py]: A good manner is to check OS install is in progress, so the code `if "Install" in workflow_obj.injectableName:` can be used here as well.

File: case/interactive/T0000_rackhd_OSInstall.py [hunk 3]
```diff
@@ -114,13 +117,21 @@ def test(self):
...
+            else:
+                self.log("WARNING", "OS installation workflow is not active on node ID: {}".format(node_id))
```
Comment (turtle-fly) [inline case/interactive/T0000_rackhd_OSInstall.py]: I think is is a result FAIL condition

File: case/interactive/T0000_rackhd_OSInstall.py (hunk 1 above)
Comment (JuneZhou19) [reply case/interactive/T0000_rackhd_OSInstall.py]: No, it's a list like [(node1_id, (node1_value1, node1_value2)),  (node2_id, (node2_value1, node2_value2)), ......] got from "self.monorail.get_nodes("compute").items()". Each time executing get_nodes function calling monorail restful for same node_id will get different tuple of node_value.
So here we can't just del by node_id like we tried before.

Comment (JuneZhou19) [reply case/interactive/T0000_rackhd_OSInstall.py]: The try exception sounds good, I can add that.

Comment (JuneZhou19) [reply case/interactive/T0000_rackhd_OSInstall.py]: Oh, seems we can't del a dic by key here, and using for loop, exception will never show here, so no need to add the try exception. And, the "node_id" pending on discovery can always be found in "node_os_install_list". :)

Comment (turtle-fly) [reply case/interactive/T0000_rackhd_OSInstall.py]: @JuneZhou19 
1. you can use `self.monorail.get_nodes("compute")` without `items()` from the very beginning to manage maps in dict and simplify code, in line 71 and 72
2. in a dict, `del(some_dict[some_key])` is ok but same sentence is not available for list
3. I agree that try... except... is not needed in this context

File: case/interactive/T0000_rackhd_OSInstall.py (hunk 2 above)
Comment (JuneZhou19) [reply case/interactive/T0000_rackhd_OSInstall.py]: I can do that.

File: case/interactive/T0000_rackhd_OSInstall.py (hunk 3 above)
Comment (JuneZhou19) [reply case/interactive/T0000_rackhd_OSInstall.py]: Other nodes' installation should also continue even the specific node fails here, and fail result of other nodes may overwrite the last failure in junit. So we choose to report a summary of all failures in the end.

Comment (turtle-fly) [reply case/interactive/T0000_rackhd_OSInstall.py]: `self.result(FAIL, msg)` will mark case as fail, and APPEND msg to the final summary, you don't have worry about overwrite.
`self.log("WARNING", msg)` here will print a warning message to log file and stdout, but nothing in final summary.

File: case/interactive/T0000_rackhd_OSInstall.py (hunk 1 above)
Comment (JuneZhou19) [reply case/interactive/T0000_rackhd_OSInstall.py]: good idea! PR submitted just now.

File: case/interactive/T0000_rackhd_OSInstall.py (hunk 3 above)
Comment (JuneZhou19) [reply case/interactive/T0000_rackhd_OSInstall.py]: OK, I see. Anyway, there is summary result for all the failed nodes of "OS installation workflow is not active" in the end now. So no worry for loosing record of installation workflow not active. Is that good?

File: case/interactive/T0000_rackhd_OSInstall.py (hunk 2 above)
Comment (JuneZhou19) [reply case/interactive/T0000_rackhd_OSInstall.py]: Included in the last PR.

**Comment by turtle-fly** [comment]
Thanks @JuneZhou19, I see necessary information get recorded~

**Pull request merged**